### ISSUE TITLE
Best AutoEncoder

### ISSUE THREAD

**Issue by kwonsik0404** [collinleiber/ClustPy#85]
Hello,

For DEC, how can I get the best AutoEncoder after the training is completed? There is one relevant attributes in the current version, saying "Final AutoEncoder". Does "Final" mean the best AutoEncoder?

Thanks.

**Comment by collinleiber** [comment]
Hi,

DEC is not trained on multiple autoencoders. Therefore, we do not receive a 'best' autoencoder, but a single autoencoder after it has been updated for a certain amount of epochs.
'Final' just indicates that the DEC training procedure has successfully been applied.

The trained autoencoder can be obtained by the following code:
dec = DEC(n_clusters=10)
dec.fit(X)
ae = dec.autoencoder

If you want to receive multiple autoencoders you have to run DEC multiple times.

**Comment by kwonsik0404** [comment]
Thanks for the clear answer.

I think that my question was not clear, so that I would like to ask you the same question but in a different way. Does the final autoencoder mean that it has optimal parameter values which result in the best representation (i.e., best features) of raw data?

Thanks in advance.

**Comment by collinleiber** [comment]
Typically, the best representation depends heavily on your task. Since DEC does not consider the reconstruction loss when optimizing the clustering result - in contrast to, e.g., IDEC -, it will generally output very compact cluster shapes.

Note that DEC is not deterministic.
Therefore, the result of DEC depends very much on the initialization (i.e., on the quality of the pretraining of the autoencoder and the initial clustering result). When running DEC multiple times, you will almost certainly get a different result with a different autoencoder (depending on the seed).

Thus, the autoencoder returned by DEC will most likely not be optimal. DEC just outputs the autoencoder present after X epochs. It can even happen that a better autoencoder with a lower loss was already identified in an earlier epoch that is not returned.

I hope the answer helps you a little.

**Comment by kwonsik0404** [comment]
Thanks for your quick response again.

I have one more question. Then, do other approaches (i.e., DKM and DCN) return an optimal autoencoder after all the training is completed?

**Comment by collinleiber** [comment]
No, the behavior of the different deep clustering approaches is highly comparable in that regard.
